### PR TITLE
Assoc item resolution refactoring + resolve `T::Item`

### PR DESCRIPTION
**Pull request by flodiebold**
This adds resolution of 'unselected' associated types for type parameters, e.g. in `foo<T: Iterator>() -> T::Item`. I actually thought rustc allowed this for any type, but as far as I can see, if you try to do it for a concrete type, it always results in 'ambiguous associated type'. So we only resolve associated types from traits in where clauses.

There's also a bunch of refactoring in preparation of resolving trait items in value NS (sorry for mixing that in here); in particular I removed the redundant and incomplete type-level resolution that was happening for the segments before the value, and made it call the type lowering code instead. Also I changed the `assoc_resolutions` to contain a new `AssocItem` enum instead of `ImplItem`s, since they will sometimes resolve to `TraitItem`s as well.

### CODE REVIEW

File: crates/ra_hir/src/ty/infer.rs [hunk 1]
```diff
@@ -503,85 +503,71 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
 
     fn resolve_assoc_item(
         &mut self,
-        mut def: TypeNs,
+        def: TypeNs,
         path: &Path,
         remaining_index: usize,
         id: ExprOrPatId,
     ) -> Option<(ValueNs, Option<Substs>)> {
```
Comment (matklad) [inline crates/ra_hir/src/ty/infer.rs]: Can we actually resolve more than once segment as assoc type? Something like this fails to compile: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff72fdd4c27f8da5443153fd33b29d93

When I've touched this code with name res refactoring, I had a feeling that some parts here are dead code

File: crates/ra_hir/src/code_model.rs [hunk 2]
```diff
@@ -1019,3 +1023,30 @@ impl Container {
...
+
+impl From<crate::ImplItem> for AssocItem {
+    fn from(i: crate::ImplItem) -> Self {
+        match i {
+            crate::ImplItem::Method(f) => AssocItem::Function(f),
+            crate::ImplItem::Const(c) => AssocItem::Const(c),
+            crate::ImplItem::TypeAlias(t) => AssocItem::TypeAlias(t),
+        }
+    }
+}
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: I wonder if there's a better encoding here than tree identical enums... Perhaps we only need `AssocItem`, and can kill trait and impl items?

File: crates/ra_hir/src/ty/lower.rs [hunk 3]
```diff
@@ -86,78 +86,128 @@ impl Ty {
...
+                    {
+                        Some(associated_ty) => {
+                            // FIXME handle type parameters on the segment
+                            Ty::Projection(ProjectionTy {
+                                associated_ty,
+                                parameters: trait_ref.substs,
+                            })
+                        }
+                        None => {
+                            // associated type not found (FIXME: report error)
```
Comment (matklad) [inline crates/ra_hir/src/ty/lower.rs]:
```suggestion
                            // FIXME: report error (associated type not found)
```

File: crates/ra_hir/src/ty/infer.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir/src/ty/infer.rs]: Ah, paths can be longer for functions:

```
let _: X::Item = X::Item::default();
```

File: crates/ra_hir/src/ty/lower.rs [hunk 4]
```diff
@@ -86,78 +86,128 @@ impl Ty {
...
+        for t in traits {
+            if let Some(associated_ty) = t.associated_type_by_name(db, &segment.name) {
+                let generics = t.generic_params(db);
+                let mut substs = Vec::new();
+                substs.push(self_ty.clone());
+                substs.extend(
+                    iter::repeat(Ty::Unknown).take(generics.count_params_including_parent() - 1),
+                );
+                // FIXME handle type parameters on the segment
+                result = Ty::Projection(ProjectionTy { associated_ty, parameters: substs.into() });
```
Comment (matklad) [inline crates/ra_hir/src/ty/lower.rs]:
```suggestion
                return Ty::Projection(ProjectionTy { associated_ty, parameters: substs.into() });
```

File: crates/ra_hir/src/ty/infer.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty/infer.rs]: Yeah, that. I think I got rid of any loops, we still have to check whether there are more segments remaining though.

File: crates/ra_hir/src/code_model.rs (hunk 2 above)
Comment (flodiebold) [reply crates/ra_hir/src/code_model.rs]: Maybe... with `TraitItem` and `ImplItem`, we could additionally ensure that they are actually in a trait resp. impl. I don't know if that's worth it though... And we could also make them simple newtypes over `AssocItem`... :thinking:

Comment (matklad) [reply crates/ra_hir/src/code_model.rs]: I'd rather just use `AssocItem` everywhere.

**Comment by flodiebold** [comment]
@matklad rebased and applied your suggestions. I did end up removing the `TypeRef` variant and hence the lifetime from the resolve function return again, because we also need to handle type-relative paths in the type context anyway.

File: crates/ra_hir/src/ty/lower.rs [hunk 5]
```diff
@@ -115,47 +115,84 @@ impl Ty {
...
+        segment: &PathSegment,
+    ) -> Ty {
+        let env = trait_env(db, resolver);
+        let traits_from_env = env.trait_predicates_for_self_ty(&self_ty).map(|tr| tr.trait_);
+        let traits = traits_from_env.flat_map(|t| t.all_super_traits(db));
+        let mut result = Ty::Unknown;
+        for t in traits {
+            if let Some(associated_ty) = t.associated_type_by_name(db, &segment.name) {
+                let generics = t.generic_params(db);
+                let mut substs = Vec::new();
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_hir/src/ty/lower.rs]: @flodiebold curious what you think here? It seems like creating a substs of a wrong length is something worth protecting against? And it seems like we always should know the target length when we create a subst?

[the remark this answers is not in the document]
Comment (flodiebold) [reply crates/ra_hir/src/ty/lower.rs]: Oh, yeah, I agree it's something we should do :slightly_smiling_face:

**Comment by matklad** [comment]
Looks like it stack overflows pretty badly, I've rolled back the commit from master

**Comment by matklad** [comment]
specifically, ` cargo run --release -p ra_cli -- analysis-stats .` fails

**Comment by bjorn3** [comment]
> specifically, cargo run --release -p ra_cli -- analysis-stats . fails

Maybe run it on CI by default.

**Comment by matklad** [comment]
I am debating this. It'll make CI much slower :( It would be a great idea to run that during release process, once we have release process :thinking:

**Pull request merged**